sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java

```diff
@@ -789,6 +799,7 @@ public CheckpointMark getCheckpointMark() {
       private boolean readerHasBeenStarted;
       private Cache<Object, UnboundedReader<OutputT>> cachedReaders;
       private Coder<UnboundedSourceRestriction<OutputT, CheckpointT>> restrictionCoder;
+      private Instant lastPreSplitWatermark;
```

Comment:
I think we can remove logics around `lastPreSplitWatermark` and update new added tests.

Reply:
Yes, you are right that it under current implementation probably plays no role. On the other hand, I'd like to avoid updating the watermark estimator to `BoundedWindow.TIMESTAMP_MAX_VALUE` (if that is not what actually came from the reader). Therefore, I'd propose to add method `isSplit` to `UnboundedSourceRestiction` and do not update the watermark estimator if that is the case (which then should result in returning `ProcessContinuation.done()`, which should be OK).

Reply:
That sounds good to me. Thanks for these changes!